gms/gossiper.cc

```diff
@@ -857,6 +857,31 @@ future<> gossiper::failure_detector_loop() {
     logger.info("failure_detector_loop: Finished main loop");
 }
 
+// Updates SUPPORTED_FEATURES application state when `feature_service::supported_feature_set()` changes.
+future<> gossiper::update_supported_features_loop() {
+    bool stop = false;
+    bool changed = false;
+    condition_variable cv;
+
+    auto subscription = _feature_service.on_supported_features_change([&] () noexcept {
+        changed = true;
+        cv.signal();
+    });
+
+    future<> watch_stopped =
+        _gossiper_stopped.wait([this] { return !_enabled; })
+        .then([&] { stop = true; cv.signal(); });
+
+    co_await [&] () -> future<> {
+        while (!stop) {
+            changed = false;
+            co_await add_local_application_state(application_state::SUPPORTED_FEATURES,
+                    versioned_value::supported_features(_feature_service.supported_feature_set()));
+            co_await cv.wait([&] { return stop || changed; });
+        }
+    }().finally([watch_stopped = std::move(watch_stopped)] () mutable { return std::move(watch_stopped); });
```

Comment:
I had to think hard on this strange lambda+finally+watch_stopped trick. I wonder if it's not possible to do this more simply with classic loops and co_await?
(but on the other hand, I can't find anything wrong with the current code either - I was just hoping it can be made a bit simpler).